File: extract_redact.py
# Goal: Given a legal transcript textblock, identify the possible issues.

import os
import json
from google import genai
from google.genai import types

# AVOID HARDCODING IN PRODUCTION
gem_key = "AIzaSyAdMIvGNU2nurZGuEuidqeoWYf0HC1Hp64"

client = genai.Client(api_key=gem_key)

system_instruction = """
You are an expert Trauma-Informed Legal Redactor. Your goal is to protect survivor anonymity.
Analyze text for 'Identity Clusters' including:
1. Direct PII (Names, SSNs, exact addresses).
2. Quasi-Identifiers: Job titles, unique physical traits, or landmarks that could re-identify someone.

For every risk found, provide:
- 'original_text': The text to hide.
- 'category': Why it's a risk (e.g., 'Quasi-Identifier').
- 'vague_ifier': A safe replacement (e.g., 'The high school' instead of 'Lincoln High').
- 'reason': Brief explanation of the risk.

Output must be ONLY a valid JSON object.
"""

def analyze_document(text_content):
    response = client.models.generate_content(
        model='gemini-3-flash-preview',
        contents=text_content,
        config=types.GenerateContentConfig(
            system_instruction=system_instruction,
            response_mime_type='application/json'
        )
    )

    try:
        return json.loads(response.text)
    except json.JSONDecodeError:
        print("Error: Model did not return valid JSON.")
        return response.text


# Read in ex.txt
file_path = './ex.txt'
with open(file_path, 'r') as file:
    sample_legal_text = file.read()
    results = analyze_document(sample_legal_text)
    print(json.dumps(results, indent=2))

File: extractRedact.js
const express = require('express');
const multer = require('multer');
const fs = require('fs');
const { GoogleGenAI } = require('@google/genai');

const app = express();
const upload = multer({ dest: 'uploads/' }); // Temporary storage for uploaded files

// 1. Setup API Client
const GEMINI_API_KEY = "apikey";
const ai = new GoogleGenAI({ apiKey: GEMINI_API_KEY });

const systemInstruction = `
You are an expert Trauma-Informed Legal Redactor. Your goal is to protect survivor anonymity by identifying sensitive data points and "Identity Clusters."

For every risk found, provide a JSON object with these exact keys:
- 'id': A unique incrementing integer starting from 0.
- 'type': The category of the data (e.g., 'name', 'location', 'job', 'date', 'contact').
- 'originalText': The specific text from the source that needs redaction.
- 'suggestedText': A "vague-ified" replacement that preserves context but removes identity (e.g., '[Witness A]' or '[a local business]').
- 'riskLevel': 'high', 'medium', or 'low' based on how easily the info could re-identify the person.
- 'riskReason': A brief explanation of why this specific text is a threat to anonymity.
- 'status': Always set this to 'pending'.

Analyze text for:
1. Direct PII: Names, specific addresses, phone numbers.
2. Quasi-Identifiers: Unique job titles (e.g., "the only female neurosurgeon"), specific landmarks, or intersectional details that, when combined, create a "mosaic effect" to identify a person.

Output must be ONLY a valid JSON array of these objects.
`;

// 2. The API Endpoint
app.post('/api/analyze', upload.single('document'), async (req, res) => {
    try {
        if (!req.file) {
            return res.status(400).json({ error: "No file uploaded. Use field name 'document'." });
        }

        // Read the text from the uploaded file
        const textContent = fs.readFileSync(req.file.path, 'utf8');

        // Call Gemini (using your logic that already works)
        const response = await ai.models.generateContent({
            model: "gemini-3-flash-preview",
            contents: [{ role: "user", parts: [{ text: textContent }] }],
            config: {
                systemInstruction: systemInstruction,
                responseMimeType: "application/json"
            }
        });

        // 3. Cleanup: Delete the temporary file from /uploads
        fs.unlinkSync(req.file.path);

        // Send JSON back to frontend
        const result = JSON.parse(response.text);
        res.json(result);

    } catch (error) {
        console.error("API Error:", error);
        // Cleanup if file exists but analysis failed
        if (req.file && fs.existsSync(req.file.path)) fs.unlinkSync(req.file.path);
        res.status(500).json({ error: "Analysis failed: " + error.message });
    }
});

// Start Server
const PORT = 3000;
app.listen(PORT, () => {
    if (!fs.existsSync('./uploads')) fs.mkdirSync('./uploads');
    console.log(`Server running on http://localhost:${PORT}`);
});